lib/rspec/core/formatters/exception_presenter.rb

```diff
@@ -0,0 +1,146 @@
+module RSpec
+  module Core
+    module Formatters
+      # @private
+      class ExceptionPresenter
+        attr_reader :exception, :example, :description, :message_color,
+                    :detail_formatter, :extra_detail_formatter, :backtrace_formatter
+        private :message_color, :detail_formatter, :extra_detail_formatter, :backtrace_formatter
+
+        def initialize(exception, example, options={})
+          @exception               = exception
+          @example                 = example
+          @message_color           = options.fetch(:message_color)          { RSpec.configuration.failure_color }
+          @description             = options.fetch(:description_formatter)  { Proc.new { example.full_description } }.call(self)
+          @detail_formatter        = options.fetch(:detail_formatter)       { Proc.new {} }
+          @extra_detail_formatter  = options.fetch(:extra_detail_formatter) { Proc.new {} }
+          @backtrace_formatter     = options.fetch(:backtrace_formatter)    { RSpec.configuration.backtrace_formatter }
+          @indentation             = options.fetch(:indentation, 2)
+          @skip_shared_group_trace = options.fetch(:skip_shared_group_trace, false)
+          @failure_lines           = options[:failure_lines]
+        end
+
+        def message_lines
+          add_shared_group_lines(failure_lines, Notifications::NullColorizer)
+        end
+
+        def colorized_message_lines(colorizer=::RSpec::Core::Formatters::ConsoleCodes)
+          add_shared_group_lines(failure_lines, colorizer).map do |line|
+            colorizer.wrap line, message_color
+          end
+        end
+
+        def formatted_backtrace
+          backtrace_formatter.format_backtrace(exception.backtrace, example.metadata)
+        end
+
+        def colorized_formatted_backtrace(colorizer=::RSpec::Core::Formatters::ConsoleCodes)
+          formatted_backtrace.map do |backtrace_info|
+            colorizer.wrap "# #{backtrace_info}", RSpec.configuration.detail_color
+          end
+        end
+
+        def fully_formatted(failure_number, colorizer=::RSpec::Core::Formatters::ConsoleCodes)
+          alignment_basis = "#{' ' * @indentation}#{failure_number}) "
+          indentation = ' ' * alignment_basis.length
+
+          "\n#{alignment_basis}#{description}#{detail_formatter.call(example, colorizer, indentation)}" \
+          "\n#{formatted_message_and_backtrace(colorizer, indentation)}" \
+          "#{extra_detail_formatter.call(failure_number, colorizer, indentation)}"
+        end
+
+        def failure_slash_error_line
+          @failure_slash_error_line ||= "Failure/Error: #{read_failed_line.strip}"
+        end
+
+      private
+
+        if String.method_defined?(:encoding)
```

Comment:
This is repeated from elsewhere isn't it? Is it worth creating something to do this rather than repeating ourselves?